subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java

```diff
@@ -122,20 +154,29 @@ public String toString() {
 
     public static class SerializerImpl extends AbstractSerializer<DefaultFileCollectionSnapshot> {
         private final SnapshotMapSerializer snapshotMapSerializer;
+        private final HashCodeSerializer hashCodeSerializer;
 
         public SerializerImpl(StringInterner stringInterner) {
             this.snapshotMapSerializer = new SnapshotMapSerializer(stringInterner);
+            this.hashCodeSerializer = new HashCodeSerializer();
         }
 
         public DefaultFileCollectionSnapshot read(Decoder decoder) throws Exception {
             TaskFilePropertyCompareStrategy compareStrategy = TaskFilePropertyCompareStrategy.values()[decoder.readSmallInt()];
+            boolean hasHash = decoder.readBoolean();
+            HashCode hash = hasHash ? hashCodeSerializer.read(decoder) : null;
             Map<String, NormalizedFileSnapshot> snapshots = snapshotMapSerializer.read(decoder);
             boolean pathIsUnique = decoder.readBoolean();
-            return new DefaultFileCollectionSnapshot(snapshots, compareStrategy, pathIsUnique);
+            return new DefaultFileCollectionSnapshot(snapshots, hash, compareStrategy, pathIsUnique);
         }
 
         public void write(Encoder encoder, DefaultFileCollectionSnapshot value) throws Exception {
             encoder.writeSmallInt(value.compareStrategy.ordinal());
+            boolean hasHash = value.getHash() != null;
```

Comment:
Shouldn't we serialize `value.hashCode` here instead?

Reply:
Good catch. That's where the performance regressions are coming from I guess. I have no clue why I didn't see this in the profiler snapshots I look at. :(